### PR TITLE
Fixes #719 add Kafka Source and HBase Sink for dsl

### LINKED ISSUE #719: add Kafka Source and HBase Sink for dsl

### PR DESCRIPTION
**Pull request by kkasravi**
Modeled after CanBuildFrom where an implicit graph is made available to each Stream[T]'s methods.  The implicit graph should be injectable (noted in TODO in StreamApp object)

[Review on Reviewable](https://reviewable.io/reviews/intel-hadoop/gearpump/834)

### CODE REVIEW

File: experiments/dsl/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala [hunk 1]
```diff
@@ -127,25 +133,38 @@ class Stream[T](dag: Graph[Op, OpEdge], private val thisNode: Op, private val ed
...
+    graph.addEdge(thisNode, edge.getOrElse(Shuffle), processorOp)
+    new Stream[R](processorOp, Some(Shuffle))
+  }
+
+  /**
+   * connect with a low level Processor(TaskDescription)
+   * @param processors
+   * @return
+   */
+  def branch(processors: ProcessorOp[_ <: Task]*)(implicit graph: Graph[Op,OpEdge]): Stream[T] = {
```
Comment (clockfly) [inline experiments/dsl/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala]: There is already process, why adding this?

File: experiments/dsl/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala [hunk 2]
```diff
@@ -23,29 +23,35 @@ import org.apache.gearpump.streaming.task.Task
 import org.apache.gearpump.util.Graph
 import org.slf4j.LoggerFactory
 
-class Stream[T](dag: Graph[Op, OpEdge], private val thisNode: Op, private val edge: Option[OpEdge] = None) {
+import scala.reflect.ClassTag
+
+class Stream[T:ClassTag](private val thisNode:Op, private val edge: Option[OpEdge] = None) {
+
+  def and: Stream[T] = {
```
Comment (clockfly) [inline experiments/dsl/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala]: ? what is this?

**Comment by clockfly** [comment]
-1

1. branch and and doesn't make sense.
2. It may not be a good idea to add an implicit graph everywhere

What I expected was something like:
streamApp.from(kafkaSource).flatmap.reduce.writeTo(hbaseSink)

File: experiments/dsl/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala (hunk 2 above)
Comment (kkasravi) [inline experiments/dsl/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala]: See PipeLineDSL 
```

    app.from(ProcessorOp(classOf[KafkaProducer], 1, "KafkaProducer")).
      branch(ProcessorOp(classOf[CpuProcessor], processors, "CpuProcessor"), ProcessorOp(classOf[CpuPersistor], persistors, "CpuPersistor")).and.
      branch(ProcessorOp(classOf[MemoryProcessor], processors, "MemoryProcessor"), ProcessorOp(classOf[MemoryPersistor], persistors, "MemoryPersistor"))
```

File: experiments/dsl/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala (hunk 1 above)
Comment (kkasravi) [inline experiments/dsl/src/main/scala/org/apache/gearpump/streaming/dsl/Stream.scala]: Stream.process does not allow branching - only linear downstream chaining. This allows multiple downstream nodes to be attached at a single upstream node.

**Comment by kkasravi** [comment]
The use case
```
streamApp.from(kafkaSource).flatmap.reduce.writeTo(hbaseSink)
```
doesn't map to what's done in PipeLine currently where the source node is branched into 2 downstream paths - one for 'cpu' the other for 'memory'. This results in 2 sinks (or writeTo's).

**Comment by kkasravi** [comment]
By the way I made an attempt to change Graph[Op,OpEdge] to Graph[V[_]<:Op,E<:OpEdge] that would allow Graph to be a typeclass so that different types of Graphs could be injected into Stream[T] that have defined new types of Op, OpEdges. This brings the design much closer to scala collections (or even scala-graph). This ends up being a pretty involved change since we pull in higher kinded types.

**Comment by kkasravi** [comment]
Investigating the build failure.

**Comment by kkasravi** [comment]
We could do something like this
```
streamApp.from(kafkaSource).flatmap(cpuprocessor).reduce(cpupersistor).to(hbaseSink).and.flatmap(cpuprocessor).reduce(cpupersistor).to(hbaseSink)

**Comment by kkasravi** [comment]
Ok, i'm removing ProcessorOp and lifting the various onNext operations out into the scope of PipeLineDSL. Right now it will only deliver batchSize messages. If this direction is a '+1', I'll continue to correctly define a source, sink Op and remove 'and', 'batch' etc.

**Comment by clockfly** [comment]
For the branch,

val root = streamApp.from(kafkaSource)
root.flatmap(cpuprocessor).reduce(cpupersistor).to(hbaseSink)
root.flatmap(cpuprocessor).reduce(cpupersistor).to(hbaseSink)

No need to use "and" or "branch"

File: experiments/dsl/src/test/scala/org/apache/gearpump/streaming/dsl/StreamAppSpec.scala [hunk 3]
```diff
@@ -39,19 +40,19 @@ class StreamAppSpec  extends FlatSpec with Matchers with BeforeAndAfterAll {
 
 
   it should "be able to generate multiple new streams" in {
-    val app = new StreamApp("dsl", context)
-    app.fromCollection(List("A").iterator, 1)
-    app.fromCollection(List("B").iterator, 1)
+    val app = StreamApp("dsl", context)
+    app(List("A"), 1)
```
Comment (clockfly) [inline experiments/dsl/src/test/scala/org/apache/gearpump/streaming/dsl/StreamAppSpec.scala]: It is not good to construct a app by a source, as one app can have multiple source.

For example:

val app = StreamApp("appName")
app.fromSource(Kafka).map.reduce.writeTo(sink)
app.from(anotherSource).writeTo(sink2)

**Comment by kkasravi** [comment]
Ok, right now i have this:
```
  def filter(dimension: String)(msg: Message): Traversable[Message] = {
    val jsonData = msg.msg.asInstanceOf[String]
    val metrics = read[Array[Datum]](jsonData)
    val data = metrics.flatMap(datum => {
      datum.dimension match {
        case dimension =>
          Some(datum)
        case _ =>
          None
      }
    }).toSeq.toArray
    Array(Message(write[Array[Datum]](data), msg.timestamp)).toTraversable
  }

  def application(context: ClientContext, config: ParseResult): Unit = {
    import Messages._
    import PipelineConfig._
    val pipeLinePath = config.getString("conf")
    val pipelineConfig = PipelineConfig(ConfigFactory.parseFile(new java.io.File(pipeLinePath)))
    val processors = pipelineConfig.getInt(PROCESSORS)
    val persistors = pipelineConfig.getInt(PERSISTORS)
    val kafkaConfig = KafkaConfig(pipelineConfig)
    val repo = new HBaseRepo {
      def getHBase(table: String, conf: Configuration): HBaseSinkInterface = HBaseSink(table, conf)
    }
    val appConfig = UserConfig.empty.withValue(KafkaConfig.NAME, kafkaConfig).withValue(PIPELINE, pipelineConfig).withValue(HBASESINK, repo)
    val source: TimeReplayableSource = new KafkaSource("KafkaSource", TaskId(0,0), 1, kafkaConfig, kafkaConfig.getMessageDecoder)
    val app = StreamApp("PipeLine", context, appConfig)
    val producer = app(source, 1, kafkaConfig.getConsumerEmitBatchSize, "kafka-producer")
    producer.flatMap(filter(CPU)).process(classOf[CpuPersistor], persistors, "cpu")
    producer.flatMap(filter(MEM)).process(classOf[MemoryPersistor], persistors, "memory")
```
Using reduce function (ReduceOp) right now doesn't do what i would like it to do, i need to create something like reduceWhile (similar to list.takeWhile) where taskContext.output is only called when (T) => Boolean returns true which in this case would be a timestamp delta.

File: experiments/dsl/src/test/scala/org/apache/gearpump/streaming/dsl/StreamAppSpec.scala (hunk 3 above)
Comment (kkasravi) [inline experiments/dsl/src/test/scala/org/apache/gearpump/streaming/dsl/StreamAppSpec.scala]: ok a new empty graph is constructed for each app instance. The example above should now result in 2 sources in the DAG. Below is a similar example:
```
val app = StreamApp("App", context, appConfig)
implicit val graph: Graph[Op,OpEdge] = app.graph
val producer1 = app(source, 1, kafkaConfig.getConsumerEmitBatchSize, "kafka-producer")
val data = "This is a good start, bingo!! bingo!!"
val producer2 = app(data.lines.toList, 1)
```
Still working on the sink/writeTo

**Comment by clockfly** [comment]
Why cannot you just use the dsl filter()?

**Comment by clockfly** [comment]
@kkasravi 
I think if we should use strong typed message, instead of doing manual serialization/deserializaton.

**Comment by kkasravi** [comment]
Yes I think i can use filter + reduce. I should have this change checked in before your Monday.

This is a good point for Graph. I'll create a case class that holds vertices and edges.

**Comment by clockfly** [comment]
Can we avoiding using implicit graph?

For example, if user want to define two StreamApp in same program, then they will conflict with each other.

**Comment by kkasravi** [comment]
I'll revert to where the Graph is defined within Stream during Stream creation. Having graph as an implicit parameter for each Stream method (similar to scala collections CanBuildFrom) doesn't quite make sense since the graph is building a DAG across Stream methods rather than helping to build the next Stream[T].

**Comment by clockfly** [comment]
@kkasravi 

It may be challenging to support any ReplayableSource in a generic way.
How about adding a OP specifially for kafka.

Something like:
```
Stream.fromKafka[T](kafkaConfig， other config) {
}
```

When we get more confidence about supporting generic Replayable source, then we can revisit this.

**Comment by kkasravi** [comment]
This latest update exposes some possible problems within the DSL - where serialization errors occur if any of the method references other methods or vals outside that method's scope. Probably a given - but suggests we handle the errors better and possibly look at spores.

**Comment by kkasravi** [comment]
I will probably need to merge master into this branch. Waiting on the build. PIpeLineDSLSpec is something to look at that is emulating kafka, hbase.

**Comment by whjiang** [comment]
Maybe we can follow Spark streaming API for Kafka:
```
import kafka.serializer.StringDecoder
import org.apache.spark.SparkConf
import org.apache.spark.streaming.{Seconds, StreamingContext}
import org.apache.spark.streaming.kafka.KafkaUtils
 
val ssc = new StreamingContext(new SparkConf, Seconds(60))
 
// hostname:port for Kafka brokers, not Zookeeper
val kafkaParams = Map("metadata.broker.list" -> "localhost:9092,anotherhost:9092")
 
val topics = Set("sometopic", "anothertopic")
 
val stream = KafkaUtils.createDirectStream[String, String, StringDecoder, StringDecoder](
  ssc, kafkaParams, topics)
```

**Comment by clockfly** [comment]
@kkasravi 

Maybe you should throw way custom serialization and try strong typed stream this time, it will work this time:)

**Comment by kkasravi** [comment]
Yes, it would make the code smaller and much more elegant. I will do so. Thanks.

**Comment by kkasravi** [comment]
I've converted PipeLineDSL and PipeLineDSLSpec to not do any custom serialization.
I'm getting a kryo serialization exception. I think I'm not entering the Array[Datum] class in the right conf file. I've added to conf/pipeline.conf

```
gearpump {
  serializers {
    "org.apache.gearpump.experiments.pipeline.Messages$Datum" = ""
    "org.apache.gearpump.experiments.pipeline.Messages.Datum" = ""
  }
}
```


```
[ERROR] [04/22/2015 13:38:00.562] [OneForOneStrategy] Encountered unregistered class ID: 257443811
com.esotericsoftware.kryo.KryoException: Encountered unregistered class ID: 257443811
        at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:119)
        at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:641)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:752)
        at org.apache.gearpump.streaming.MessageSerializer.read(MessageSerializer.scala:34)
        at org.apache.gearpump.streaming.MessageSerializer.read(MessageSerializer.scala:26)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:764)
        at org.apache.gearpump.serializer.FastKryoSerializer.deserialize(FastKryoSerializer.scala:66)
        at org.apache.gearpump.transport.netty.Server$$anonfun$msgHandler$1$$anonfun$applyOrElse$3$$anonfun$apply$1.apply(Server.scala:70)
        at org.apache.gearpump.transport.netty.Server$$anonfun$msgHandler$1$$anonfun$applyOrElse$3$$anonfun$apply$1.apply(Server.scala:67)
        at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at org.apache.gearpump.transport.netty.Server$$anonfun$msgHandler$1$$anonfun$applyOrElse$3.apply(Server.scala:67)
        at org.apache.gearpump.transport.netty.Server$$anonfun$msgHandler$1$$anonfun$applyOrElse$3.apply(Server.scala:61)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
        at scala.collection.immutable.Map$Map1.foreach(Map.scala:116)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
        at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        at org.apache.gearpump.transport.netty.Server$$anonfun$msgHandler$1.applyOrElse(Server.scala:61)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:170)
        at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
        at org.apache.gearpump.transport.netty.Server.aroundReceive(Server.scala:34)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
        at akka.actor.ActorCell.invoke(ActorCell.scala:487)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:254)
        at akka.dispatch.Mailbox.run(Mailbox.scala:221)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:231)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by clockfly** [comment]
@kkasravi 

Please search the LOG for 257443811, there should be some log saying implicit register ... class 
Then, you need to add that class into the serializer list.

**Comment by kkasravi** [comment]
Ok, thanks. The PipeLineDSL example is looking pretty compelling. I think there is still some amount of refinement for a standard Source and Sink but it is getting there.

**Comment by clockfly** [comment]
@kkasravi 

For the kryo exception, I can reproduce it. There is no information about which class cannot be serialized. 

I have created a issue https://github.com/intel-hadoop/gearpump/issues/871 
And will submit a PR soon.

**Comment by clockfly** [comment]
@kkasravi 

https://github.com/intel-hadoop/gearpump/pull/872 should solve the problem.
When there is error log, you then need to add the class name into gear.conf serializer list.

**Comment by kkasravi** [comment]
Thanks @clockfly. I'm now testing with Kafka and HBase on a hadoop cluster. I've verified that it works for stubbed/mock interfaces. I'll update when the kafka/hbase pathway is working. If you're anxious to merge this to master, i can get this working on new issue.

**Comment by clockfly** [comment]
@kkasravi 

Generally, I think the patch is good.

Is it possible to add a shortcut for kafka and hbase. For example, add a operator in Stream called fromKafka, and add a sink in Stream called writeToHbase()?

Then the PipleLineDSL code will be much simplified. 

Currently, I see in HbaseSink, there is also many map-like transformation job, is it possible to break that to a map operator and a pure-simple hbase sink?

**Comment by kkasravi** [comment]
@clockfly Yes, I'll add these operators. See #53 for recent updates made today and some related questions.

**Comment by kkasravi** [comment]
Hi @clockfly the PipeLineDSL and PipeLineDSLSpec are much shorter 
```
    val app = StreamApp("PipeLineDSL", context, appConfig)
    val producer = app.readFromKafka(new ReplayableSource(kafkaConfig), 1, "time-replayable-producer")

    producer.flatMap(metrics => {
      Some(metrics.flatMap(datum => {
        datum.dimension match {
          case CPU =>
            Some(datum)
          case _ =>
            None
        }
      }))
    }).reduce((msg1, msg2) => {
      val now = System.currentTimeMillis
      def average(average: TAverage)(metrics: Array[Datum]): Array[Datum] = {
        metrics.flatMap(datum => {
          average.average(datum, now)
        })
      }
      average(TAverage(50))(msg2)
    }).writeToHBase(toHBaseSink)

    producer.flatMap(metrics => {
      Some(metrics.flatMap(datum => {
        datum.dimension match {
          case MEM =>
            Some(datum)
          case _ =>
            None
        }
      }))
    }).reduce((msg1, msg2) => {
      val now = System.currentTimeMillis
      def average(average: TAverage)(metrics: Array[Datum]): Array[Datum] = {
        metrics.flatMap(datum => {
          average.average(datum, now)
        })
      }
      average(TAverage(100))(msg2)
    }).writeToHBase(toHBaseSink)
```

File: experiments/kafka-hbase-pipeline/src/main/scala/org/apache/gearpump/experiments/pipeline/PipeLineDSL.scala [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    val persistors = pipeLineConfig.getInt(PERSISTORS)
+    val kafkaConfig = KafkaConfig(pipeLineConfig)
+    val repo = new HBaseRepo {
+      def getHBase(table: String, conf: Configuration): HBaseSinkInterface = HBaseSink(table, conf)
+    }
+    val appConfig = UserConfig.empty.withValue(KafkaConfig.NAME, kafkaConfig).withValue(PIPELINE, pipeLineConfig).withValue(HBASESINK, repo)
+    System.setProperty(Constants.GEARPUMP_CUSTOM_CONFIG_FILE, pipeLinePath)
+
+    val app = StreamApp("PipeLineDSL", context, appConfig)
+    val producer = app.readFromKafka(new ReplayableSource(kafkaConfig), 1, "time-replayable-producer")
```
Comment (clockfly) [inline experiments/kafka-hbase-pipeline/src/main/scala/org/apache/gearpump/experiments/pipeline/PipeLineDSL.scala]: Can this be simplified by only relying on a kakfaConfig?

For example:
app.readfromKafka(kafkaConfig, name)

**Comment by kkasravi** [comment]
app.fromKafka now takes a kafkaConfig and a message converter to deserialize the message to the target type. 
```
    val producer = app.readFromKafka(kafkaConfig, msg => {
      val jsonData = msg.msg.asInstanceOf[String]
      val envelope = read[Envelope](jsonData)
      val body = read[Body](envelope.body)
      body.metrics
    }, 1, "time-replayable-producer")
```
I think the converter (Message => T) is unavoidable since the user may or may not have control over the kafka topic schema

**Comment by clockfly** [comment]
+1

**Comment by clockfly** [comment]
committed.

**Pull request closed without merging**